### PR TITLE
feat(rs-dpp): introduce `StateTransitionFactory`

### PR DESCRIPTION
**Pull request by jawid-h**
## Issue being fixed or feature implemented

Introducing `StateTransitionFactory` for both `rs-dpp` and `wasm-dpp`

## What was done?

- added `StateTransitionFactoy` structure in `rs-dpp`
- added `StateTransitionBasicValidator` structure to join different state transition basic validators into one
- added `StateTransitionFactory` structure and bindings to `wasm-dpp`

## How Has This Been Tested?

- unit tests

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/rs-dpp/src/state_transition/state_transition_factory.rs [hunk 1]
```diff
@@ -19,15 +26,103 @@ use crate::{
...
+
+    pub async fn create_from_object(
+        &self,
+        raw_state_transition: JsonValue,
+        options: Option<StateTransitionFactoryOptions>,
+    ) -> Result<StateTransition, ProtocolError> {
+        let options = options.unwrap_or_default();
+
+        if !options.skip_validation {
+            let execution_context = StateTransitionExecutionContext::default();
```
Comment (antouhou) [inline packages/rs-dpp/src/state_transition/state_transition_factory.rs:81]: Do we always want the context to be the default context here?

File: packages/wasm-dpp/src/state_transition/errors/invalid_state_transition_error.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+use dpp::consensus::ConsensusError;
+use wasm_bindgen::{prelude::wasm_bindgen, JsValue};
+
+use crate::errors::consensus_error::from_consensus_error_ref;
+
+#[wasm_bindgen(js_name = InvalidStateTransitionError)]
+pub struct InvalidStateTransitionError {
```
Comment (antouhou) [inline packages/wasm-dpp/src/state_transition/errors/invalid_state_transition_error.rs:7]: I recommend to rename this struct to `InvalidStateTransitionErrorWasm` for clarity, as with all other errors

File: packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js [hunk 3]
```diff
@@ -1,160 +1,99 @@
...
-    decodeProtocolEntityMock = this.sinonSandbox.stub();
-
-    validateStateTransitionBasicMock = this.sinonSandbox.stub();
-    createStateTransitionMock = this.sinonSandbox.stub().returns(stateTransition);
-
-    dppMock = createDPPMock();
+    stateRepositoryMock = createStateRepositoryMock(this.sinonSandbox);
+
+    stateRepositoryMock.fetchDataContract.resolves(dataContract);
+    stateRepositoryMock.createDataContract.resolves();
```
Comment (antouhou) [inline packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js:43]: Do we need to resolve to empty functions everywhere? I think mock by default is kinda doing that already

File: packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js [hunk 4]
```diff
@@ -1,160 +1,99 @@
...
 
-      expect(result).to.equal(stateTransition);
-
-      expect(validateStateTransitionBasicMock).to.have.not.been.called();
-
-      expect(createStateTransitionMock).to.have.been.calledOnceWith(rawStateTransition);
-    });
-
-    it('should throw InvalidStateTransitionError if passed object is not valid', async () => {
-      const validationError = new SomeConsensusError('test');
```
Comment (antouhou) [inline packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js:72]: Can you just create a test where `InvalidStateTransitionError` is thrown? Without any mocks and stuff

File: packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js [hunk 5]
```diff
@@ -1,160 +1,99 @@
...
-
-      expect(factory.createFromObject).to.have.been.calledOnceWith(rawStateTransition);
-
-      expect(decodeProtocolEntityMock).to.have.been.calledOnceWith(
-        serializedStateTransition,
-      );
-    });
-
-    it('should throw InvalidStateTransitionError if the decoding fails with consensus error', async () => {
-      const parsingError = new SerializedObjectParsingError(
```
Comment (antouhou) [inline packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js:127]: Same as above

File: packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js (hunk 3 above)
Comment (markin-io) [reply packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js:43]: @shumkov requested to get rid of default `resolve()` a while ago. They have to be provided explicitly

File: packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js (hunk 4 above)
Comment (jawid-h) [reply packages/wasm-dpp/test/unit/stateTransition/StateTransitionFactory.spec.js:72]: I'll try

File: packages/wasm-dpp/src/state_transition/errors/invalid_state_transition_error.rs (hunk 2 above)
Comment (jawid-h) [reply packages/wasm-dpp/src/state_transition/errors/invalid_state_transition_error.rs:7]: Ok

File: packages/rs-dpp/src/state_transition/state_transition_factory.rs (hunk 1 above)
Comment (jawid-h) [reply packages/rs-dpp/src/state_transition/state_transition_factory.rs:81]: I did as in other factories we have, so I'm not sure myself

Comment (antouhou) [reply packages/rs-dpp/src/state_transition/state_transition_factory.rs:81]: The original implementation does indeed have default context, thank you!

**Review by antouhou — APPROVED** [review]
👍

**Pull request merged**